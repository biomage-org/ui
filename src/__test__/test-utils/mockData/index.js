import generateMockExperimentData from './mockExperimentData';
import generateMockProjects from './mockProjects';
import generateMockSamples from './mockSamples';
import generateMockExperiment from './mockExperiment';

// This file is a work in progress to generate mock data for the endpoints of the test.
// The generated data should be usable as response data to be sent via the mocked API in mockAPI.js
// The module should also expose the generated data that can be imported and used easily in tests
// A ticket has been created to give more thought into this https://biomage.atlassian.net/browse/BIOMAGE-1553

const responseData = {};

const projects = generateMockProjects(2);
responseData.projects = projects;

const projectWithSamples = responseData.projects[0];

responseData.samples = [generateMockSamples(
  projectWithSamples.uuid,
  projectWithSamples.experiments[0],
  3,
)];

const { samples } = responseData.samples[0];
const sampleIds = Object.keys(samples);

projectWithSamples.samples = sampleIds;

responseData.experiments = [generateMockExperiment(
  projectWithSamples.uuid,
  projectWithSamples.experiments[0],
  projectWithSamples.experiments[0],
  projectWithSamples.samples,
)];

responseData.experimentData = generateMockExperimentData(
  projectWithSamples.uuid,
  projectWithSamples.experiments[0],
  projectWithSamples.experiments[0],
  projectWithSamples.samples,
);

export {
  projects,
  samples,
  responseData,
};
