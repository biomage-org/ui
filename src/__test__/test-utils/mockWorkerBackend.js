import * as socketConnectionMocks from 'utils/socketConnection';
import * as seekWorkResponseMocks from 'utils/work/dispatchWorkRequest';

import SocketMock from 'socket.io-mock';

const socketMock = new SocketMock();

jest.mock('utils/work/seekWorkResponse', () => {
  const mockSeekFromS3 = jest.fn();
  const originalModule = jest.requireActual('utils/work/seekWorkResponse');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    seekFromS3: mockSeekFromS3,
  };
});

jest.mock('aws-amplify', () => ({
  configure: () => ({
    Storage: {
      AWSS3: {
        bucket: 'biomage-originals-test',
      },
    },
  }),
  Storage: {
    get: async (ETag) => `http://mock.s3.amazonaws.com/${ETag}`,
  },
  Auth: {
    federatedSignIn: jest.fn(),
    signOut: jest.fn(),
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
    currentSession: () => ({
      getIdToken: (() => ({
        getJwtToken: () => 'fakeJwtToken',
      })),
    }),
  },
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();
  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '1234' });
    }),
    mockEmit,
    mockOn,
  };
});

// To mock worker response, modify the response returned from the API
// Set to null to force fetch from API
seekWorkResponseMocks.seekFromS3.mockImplementation(() => null);

// Set up socket emitter mock
socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
  const responseBody = { response: { error: false } };

  // After emitting, send reply to listener
  socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, responseBody);
});

// Set up socket listener mock
socketConnectionMocks.mockOn.mockImplementation((channel, f) => {
  socketMock.on(channel, (responseBody) => f(responseBody));
});
