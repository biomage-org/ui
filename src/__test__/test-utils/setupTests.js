import '@testing-library/jest-dom/extend-expect';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import preloadAll from 'jest-next-dynamic';

configure({ adapter: new Adapter() });
jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
