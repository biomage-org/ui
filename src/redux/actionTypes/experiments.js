const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

const EXPERIMENTS_SWITCH = `${EXPERIMENTS}/switch`;

/**
 * Set active project.
 */
const PROJECTS_SET_ACTIVE = `${EXPERIMENTS}/setActive`;

/**
 * Add project metadata.
 */
const PROJECTS_METADATA_CREATE = `${EXPERIMENTS}/metadataCreate`;

/**
 * Update project metadata.
 */
const PROJECTS_METADATA_UPDATE = `${EXPERIMENTS}/metadataUpdate`;

/**
 * Delete project metadta.
 */
const PROJECTS_METADATA_DELETE = `${EXPERIMENTS}/metadataDelete`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SWITCH,
  PROJECTS_SET_ACTIVE,
  PROJECTS_METADATA_CREATE,
  PROJECTS_METADATA_UPDATE,
  PROJECTS_METADATA_DELETE,
};
