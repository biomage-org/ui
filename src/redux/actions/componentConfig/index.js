import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import fetchDotPlotData from 'redux/actions/componentConfig/fetchDotPlotData';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  fetchDotPlotData,
  updatePlotData,
  resetPlotConfig,
};
