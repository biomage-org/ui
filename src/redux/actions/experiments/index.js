import createExperiment from './createExperiment';
import updateExperiment from './updateExperiment';
import loadExperiments from './loadExperiments';
import loadExperiment from './loadExperiment';
import switchExperiment from './switchExperiment';
import reorderSamples from './reorderSamples';

export {
  createExperiment,
  updateExperiment,
  loadExperiments,
  loadExperiment,
  switchExperiment,
  reorderSamples,
};
