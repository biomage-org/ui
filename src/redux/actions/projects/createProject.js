import moment from 'moment';

import {
  PROJECTS_CREATE,
  PROJECTS_ERROR,
  PROJECTS_SAVING,
} from 'redux/actionTypes/projects';

import { projectTemplate } from 'redux/reducers/projects/initialState';
import createExperiment from 'redux/actions/experiments/createExperiment';
import endUserMessages from 'utils/endUserMessages';
import handleError from 'utils/http/handleError';

const createProject = (
  projectName,
  projectDescription,
  newExperimentName,
) => async (dispatch) => {
  const createdDate = moment().toISOString();

  try {
    dispatch({
      type: PROJECTS_SAVING,
      payload: {
        message: endUserMessages.SAVING_PROJECT,
      },
    });

    const { id: experimentId } = await dispatch(createExperiment(newExperimentName));

    const newProject = {
      ...projectTemplate,
      name: projectName,
      description: projectDescription,
      uuid: experimentId,
      experiments: [experimentId],
      createdDate,
      lastModified: createdDate,
    };

    dispatch({
      type: PROJECTS_CREATE,
      payload: { project: newProject },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_CREATING_PROJECT);

    dispatch({
      type: PROJECTS_ERROR,
      payload: {
        error: errorMessage,
      },
    });
  }
};

export default createProject;
