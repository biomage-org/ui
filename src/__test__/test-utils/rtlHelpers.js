import { fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

// dropFilesIntoDropzone takes an input element and simulates droping files into it
// inputElement: this needs to be the actual input element not the dropzone itself, e.g.:
// * <input data-testid='drop-input'/> => screen.getByTestId('drop-input');
// files should be a list of files, e.g.:
// * new File(['content'], 'file.txt', { type: 'text/plain' });
const dropFilesIntoDropzone = async (inputElement, files) => {
  await act(async () => {
    fireEvent.drop(inputElement, {
      target: { files },
    });
  });
};

export {
  // eslint-disable-next-line import/prefer-default-export
  dropFilesIntoDropzone,
};
