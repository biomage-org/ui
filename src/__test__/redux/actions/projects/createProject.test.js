import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import handleError from 'utils/http/handleError';
import createProject from 'redux/actions/projects/createProject';
import initialProjectsState from 'redux/reducers/projects';

import { createExperiment } from 'redux/actions/experiments';
import {
  PROJECTS_CREATE, PROJECTS_SAVING, PROJECTS_ERROR,
} from 'redux/actionTypes/projects';
import '__test__/test-utils/setupTests';

jest.mock('utils/http/handleError');

const mockStore = configureStore([thunk]);

const experimentId = 'random-experiment-uuid';
jest.mock('redux/actions/experiments/createExperiment');
createExperiment.mockImplementation(() => async () => (experimentId));

enableFetchMocks();

describe('createProject action', () => {
  let store;

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());

    fetchMock.resetMocks();
    fetchMock.doMock();

    store = mockStore({
      projects: initialProjectsState,
    });
  });

  const projectName = 'test project';
  const projectDescription = 'test project description';
  const experimentName = 'mockExperimentName';

  it('Creates a project when there are no errors', async () => {
    fetchMock.mockResponse(JSON.stringify({}));

    await store.dispatch(
      createProject(projectName, projectDescription, experimentName),
    );

    expect(createExperiment).toHaveBeenCalledWith(experimentName, projectDescription);

    // Sends correct actions
    const actions = store.getActions();
    expect(actions[0].type).toEqual(PROJECTS_SAVING);
    expect(actions[1].type).toEqual(PROJECTS_CREATE);

    // No other action was sent
    expect(actions).toHaveLength(2);

    expect(actions[1].payload).toMatchSnapshot();
  });

  it('Shows error when there was an experiment error', async () => {
    const fetchErrorMessage = 'some error';

    createExperiment.mockImplementationOnce(() => { throw new Error(fetchErrorMessage); });

    await store.dispatch(
      createProject(projectName, projectDescription, experimentName),
    );

    expect(createExperiment).toHaveBeenCalledWith(experimentName, projectDescription);

    // Sends correct actions
    const actions = store.getActions();

    expect(actions[0].type).toEqual(PROJECTS_SAVING);
    expect(actions[1].type).toEqual(PROJECTS_ERROR);

    // Check no other action was sent
    expect(actions).toHaveLength(2);

    expect(handleError).toHaveBeenCalledTimes(1);
  });
});
