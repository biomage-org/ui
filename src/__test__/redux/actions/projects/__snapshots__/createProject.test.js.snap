// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createProject action Still creates project when using v2 1`] = `
Object {
  "project": Object {
    "createdDate": "2020-01-01T00:00:00.000Z",
    "description": "test project description",
    "experiments": Array [
      "random-experiment-uuid",
    ],
    "lastAnalyzed": null,
    "lastModified": "2020-01-01T00:00:00.000Z",
    "metadataKeys": Array [],
    "name": "test project",
    "samples": Array [],
    "uuid": "random-experiment-uuid",
  },
}
`;

exports[`createProject action Works correctly when there are no errors 1`] = `
Object {
  "createdDate": "2020-01-01T00:00:00.000Z",
  "description": "test project description",
  "experiments": Array [
    "random-experiment-uuid",
  ],
  "lastAnalyzed": null,
  "lastModified": "2020-01-01T00:00:00.000Z",
  "metadataKeys": Array [],
  "name": "test project",
  "samples": Array [],
  "uuid": "random-project-uuid",
}
`;

exports[`createProject action Works correctly when there are no errors 2`] = `
Object {
  "project": Object {
    "createdDate": "2020-01-01T00:00:00.000Z",
    "description": "test project description",
    "experiments": Array [
      "random-experiment-uuid",
    ],
    "lastAnalyzed": null,
    "lastModified": "2020-01-01T00:00:00.000Z",
    "metadataKeys": Array [],
    "name": "test project",
    "samples": Array [],
    "uuid": "random-project-uuid",
  },
}
`;
